apps/xclock/man/xclock.man
.\" $Xorg: xclock.man,v 1.4 2001/02/09 02:05:39 xorgcvs Exp $
.\" Copyright 1988, 1994, 1998  The Open Group
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" Except as contained in this notice, the name of The Open Group shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from The Open Group.
.\"
.\" $XFree86: xc/programs/xclock/xclock.man,v 1.15 2003/07/04 16:24:30 eich Exp $
.\"
.TH XCLOCK 1 __xorgversion__
.SH NAME
xclock \- analog / digital clock for X
.SH SYNOPSIS
\fBxclock\fP [ \fB\-help\fP ] [ \fB\-analog\fP | \fB\-digital\fP ]
[ \fB\-brief\fP ] [ \fB\-chime\fP ] [ \fB\-hd\fP \fIcolor\fP ]
[ \fB\-hl\fP \fIcolor\fP ] [ \fB\-update\fP \fIseconds\fP ]
[ \fB\-strftime\fP \fIformat\fP ]
[ \fB\-twelve\fP | \fB-twentyfour\fP | \fB-utime\fP ]
[ \fB\-padding\fP \fInumber\fP ]
[ \fB\-norender\fP ]
[ \fB\-render\fP ]
[ \fB\-sharp\fP ]
[ \fB\-face\fP \fIpattern\fP ]
.SH DESCRIPTION
The
.I xclock
program displays the time in analog or digital form.  The time is continuously
updated at a frequency which may be specified by the user.
.SH OPTIONS
.I Xclock
accepts all of the standard X Toolkit command line options along with the
additional options listed below:
.TP 8
.B \-help
This option indicates that a brief summary of the allowed options should be
printed on the standard error.
.TP 8
.B \-analog
This option indicates that a conventional 12 hour clock face with tick marks
and hands should be used.  This is the default.
.TP 8
\fB\-digital\fP or \fB\-d\fP
This option indicates that a 24 hour digital clock should be used.
.TP 8
\fB\-brief\fP
This option indicates that the digital clock should only display the hours
and minutes fields.  The default is to show the full time and date information.
.TP 8
\fB\-utime\fP
This option indicates that a digital clock should display seconds since
the Epoch (in format '970012340 seconds since Epoch' instead of a standard
24-hour time.
.TP 8
.B \-strftime \fIformat\fP
This option allows an strftime(__libmansuffix__) format string to be specified
for the digital clock's display.
.TP 8
.B \-twelve
This option indicates that a digital clock should display the time in
twelve hour format.
This is only effective for the brief display.
.TP 8
.B \-twentyfour
This option indicates that a digital clock should display the time in
twenty-four hour format.
This is only effective for the brief display.
This is the default when a digital clock is used.
.TP 8
.B \-chime
This option indicates that the clock should chime
once on the half hour and twice on the hour.
.TP 8
\fB\-hands\fP \fIcolor\fP (or \fB\-hd\fP \fIcolor\fP)
This option specifies the color of the hands on an analog clock.  The default
is \fIblack\fP.
This option is effectively ignored when Xrender is in use.
.TP 8
\fB\-highlight\fP \fIcolor\fP (or \fB\-hl\fP \fIcolor\fP)
This option specifies the color of the edges of the hands on an analog clock,
and is only useful on color displays.  The default is \fIblack\fP.
This option is effectively ignored when Xrender is in use.
.TP 8
.B \-update \fIseconds\fP
This option specifies the frequency in seconds at which \fIxclock\fP
should update its display.  If the clock is obscured and then exposed,
it will be updated immediately.  A value of 30 seconds or less will enable a
second hand on an analog clock.  The default is 60 seconds.
.TP 8
.B \-padding \fInumber\fP
This option specifies the width in pixels of the padding
between the window border and clock text or picture.  The default is 10
on a digital clock and 8 on an analog clock.
.TP 8
.B \-render
This option tells \fIxclock\fP to use the Xrender extension to draw an
anti-aliased face. This is the default if \fIxclock\fP has been
compiled with Xrender support.
Note that the color selection options and resources used when Xrender is
in effect differ from the standard options.
.TP 8
.B \-norender
This option turns off the use of Xrender to draw the clock.
.TP 8
.B \-sharp
This option tells \fIxclock\fP to use sharper edges when drawn using
the Xrender extension.
.TP 8
.B \-face \fIpattern\fP
This option specifies the font to use in digital mode when the
Xrender extension is used.  Patterns are specified using the
fontconfig face format described in the
.I Font Names
section of
.IR fonts.conf (__filemansuffix__).
.SH X DEFAULTS
This program uses the
.I Clock
widget.  It understands all of the core resource names and classes as well as:
.PP
.TP 8
.B width (\fPclass\fB Width)
Specifies the width of the clock.  The default for analog clocks is 164
pixels; the default for digital clocks is whatever is needed to hold the
clock when displayed in the chosen font.
.TP 8
.B height (\fPclass\fB Height)
Specifies the height of the clock.  The default for analog clocks is 164
pixels; the default for digital clocks is whatever is needed to hold the
clock when displayed in the chosen font.
.TP 8
.B update (\fPclass\fB Interval)
Specifies the frequency in seconds at which the time should be redisplayed.
.TP 8
.B foreground (\fPclass\fB Foreground)
Specifies the color for the tick marks. The default depends on whether
\fIreverseVideo\fP is specified.  If \fIreverseVideo\fP is specified
the default is \fIlwhite\fP, otherwise the default is \fIblack\fP.

.TP 8
.B hands (\fPclass\fB Foreground)
Specifies the color of the insides of the clock's hands. The default
depends on whether
\fIreverseVideo\fP is specified.  If \fIreverseVideo\fP is specified
the default is \fIlwhite\fP, otherwise the default is \fIblack\fP.
Note that this resource is not used when Xrender is in effect.
.TP 8
.B highlight (\fPclass\fB Foreground)
Specifies the color used to highlight the clock's hands. The default is
 depends on whether
\fIreverseVideo\fP is specified.  If \fIreverseVideo\fP is specified
the default is \fIlwhite\fP, otherwise the default is \fIblack\fP.
Note that this resource is not used when Xrender is in effect.
.TP 8
.B analog (\fPclass\fB Boolean)
Specifies whether or not an analog clock should be used instead of a digital
one.  The default is True.
.TP 8
.B twentyfour (\fPclass\fB Boolean)
Specifies whether or not a digital clock should display the time in
twenty-four hour format.  The default is True.
.TP 8
.B chime (\fPclass\fB Boolean)
Specifies whether or not a bell should be rung on the hour and half hour.
.TP 8
.B padding (\fPclass\fB Margin)
Specifies the amount of internal padding in pixels to be used.  The default is
8.
.TP 8
.B font (\fPclass\fB Font)
Specifies the font to be used for the digital clock.  Note that variable width
fonts currently will not always display correctly.
This font is only used when Xrender is not in effect.
.TP 8
.B render (\fPclass\fB Boolean)
Specifies whether or not the Xrender extension should be used for the
display. The default is True if \fIxclock\fP has been compiled with
Xrender support.
.PP
When Xrender is in effect, the following additional resources are
understood:
.PP
.TP 8
.B face (\fPclass\fB FaceName)
Specify the pattern for the font to be used for the digital clock when
Xrender is used.   Patterns are specified using the fontconfig face format
described in the
.I Font Names
section of
.IR fonts.conf (__filemansuffix__).
.TP 8
.B sharp (\fPclass\fB Boolean)
Specifies if sharp edges should be used when rendering the clock.
The default is False.
.TP 8
.B buffer (\fPclass\fB Boolean)
Specifies that the updates of the image are drawn to a pixmap before copied
into the window instead drawing them into the window directly.
.PP
The defaults of the following color resources depend on whether
\fIreverseVideo\fP is specified.
If \fIreverseVideo\fP is specified the default is \fIlwhite\fP,
otherwise the default is \fIblack\fP.
.TP 8
.B hourColor (\fPclass\fB Foreground)
The color of the hour hand.
.TP 8
.B minuteColor (\fPclass\fB Foreground)
The color of the minute hand.
.TP 8
.B secondColor (\fPclass\fB Foreground)
The color of the second hand.
.TP 8
.B majorColor (\fPclass\fB Foreground)
The color of the major scale ticks (i. e. each five minutes).
.TP 8
.B minorColor (\fPclass\fB Foreground)
The color of the minor scale ticks (between major ticks).
.SH WIDGETS
In order to specify resources, it is useful to know the hierarchy of
the widgets which compose \fIxclock\fR.  In the notation below,
indentation indicates hierarchical structure.  The widget class name
is given first, followed by the widget instance name.
.sp
.ft CW
.nf
XClock  xclock
        Clock  clock
.fi
.ft
.sp
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH FILES
.TP
.I __apploaddir__/XClock
specifies required resources
.SH "SEE ALSO"
X(__miscmansuffix__), xrdb(__appmansuffix__), time(__libmansuffix__), fonts.conf (__filemansuffix__).
.SH BUGS
.I Xclock
believes the system clock.
.PP
When in digital mode, the string should be centered automatically.
.SH AUTHORS
Tony Della Fera (MIT-Athena, DEC)
.br
Dave Mankins (MIT-Athena, BBN)
.br
Ed Moy (UC Berkeley)



libX11/NEWS
Version 1.1.2 - 2007-06-03

* Security fixes:
  * Fix CVE-2007-1667: Multiple integer overflows in the XGetPixel() and
    XInitImage() functions.
  * Bug #9279: Fixed a file descriptor leak.

* Xlib/XCB: Ensure _XReadEvents reads at least one new event and blocks for
  exactly one event.  Large performance improvement for XIfEvent and similar.

* Hold XCB's Xlib lock even when only the user lock (XLockDisplay) is held.
  An Xlib client can query Display state, such as with NextRequest, while it
  holds only the Xlib user lock (between XLockDisplay and XUnlockDisplay), so
  XCB requests in other threads should be blocked when the Xlib user lock is
  held.  We acquire the lock even when XInitThreads was not called, so that
  pure XCB code can use multiple threads even in an otherwise single-threaded
  Xlib application.

* Allow re-entrant Xlib calls from _XIOError.
  Some libraries try to clean up X resources from atexit handlers, _fini, or
  C++ destructors. To make these work, downgrade the Display lock to a user
  lock (as in XLockDisplay) before calling exit(3).  This blocks Xlib calls
  from threads other than the one calling exit(3) while still allowing the
  exit handlers to call Xlib.

* Xlib/XCB: Avoid re-crashing after _XIOError.

* Xlib/XCB: Only remove pending_requests when there are provably no more responses.

* For NetBSD, define XTHREADLIB and XTHREAD_CFLAGS.

* Bug #7703: Fixed XSetSizeHints() et al wrt use of uninitialized data.
  Now only those fields of the respective hint struct are set that
  are actually valid in the input data.

* Bug #10292: Fixed a memory leak related to XOpenDisplay() in the XCB code.

* Bug #7713: Initialize all of the event's fields before sending it.

* Bug #10562: Define _GNU_SOURCE on glibc systems.

* Use unistd.h to get getresuid() and friends, now that we have _GNU_SOURCE
  defined.

* Switched function definitions from K&R to ANSI style.

* Many constness fixes.

* Fixed a few warnings.

* Fix 64bit issues with reallocation.

* Manpage fixes:
  * Add man pages for XKB APIs.
    (Volunteer needed to convert prototypes in man pages to ANSI C style...)
  * Protect C comments and #defines in XKB man pages from being mangled by cpp
  * Fix typo in nroff macro in XkbAddGeomOverlayKey.man
  * Bug #9695: Fixed a few argument types in the XOpenIM manpage.
  * Markup tweak for XOpenIM.
  * Bug #9696: refer to XDefineCursor() instead of XDefineCusor().
  * Bug #9697: Fixed documentation of XVisualInfo struct.  The "depth" member
    was said to be unsigned int, but it's signed.

* nls fixes:
  * Update el_GR.UTF-8/Compose.pre to match changes in xkeyboard-config cvs.
  * For nls/*.pre, allow people to comment lines by starting them with '##'.

* Build system fixes:
  * Add autogen.sh to EXTRA_DIST.
  * Makefile.am: make ChangeLog hook safer
  * Don't distribute the configure-generated XlibConf.h in tarballs.

Version 1.1.1 - 2006-11-30

The "xcompmgr is hard, let's go releasing" release

* Fix Bug #9154 and friends, to unbreak xcompmgr:
  Always process an event for _XReadEvents, even if an error occurs.
  Previously, process_responses (in the wait_for_first_event case called
  from _XReadEvents) considered any return from xcb_wait_for_event
  sufficient to think it had processed an event. If xcb_wait_for_event
  returned an error, and no more events occurred before process_responses
  called xcb_poll_for_event, process_responses would try to return with
  dpy->head NULL, and would fail an assertion for the _XReadEvents
  postcondition. Now, process_responses continues using xcb_wait_for_event
  until it gets an event.

* Bug #9153: Fix access to freed memory.
  The fix for bug #8622 introduced a smaller bug where _XReply would
  read memory shortly after freeing it. This patch caches the needed
  value in a stack-allocated variable before the heap-allocated memory
  is freed.

* libx11 doesn't use inputproto in public headers; don't require it in x11.pc.
  Based on a Debian patch.

* Debian bug #354315: Clarify return value in XGetWindowAttributes man page.
  Patch by Debian user Ross Combs.


Version 1.1 - 2006-11-23

* Add note in man-page that XListFontsWithInfo is not thread-safe.  _XReply
  drops the Display lock, so the value of dpy->request may change before
  _XReply is called again.  Jamey Sharp discovered this by inspection a few
  years ago.

* Fix Bug #8622, by fixing the response processing order for threaded apps.
  process_responses (the common code for _XReply, _XReadEvents, and
  _XEventsQueued) now handles responses in order, by adding condition variables
  to the list of outstanding requests in dpy->xcb->pending_requests, and
  blocking on them when those requests should get processed, to allow _XReply
  to process them; if actually called from _XReply, it returns when _XReply's
  request should get processed.  _XReply broadcasts on its condition variable
  after it has read its reply and re-acquired the display lock.

* Don't hold the display lock around callbacks to the application. This avoids
  recursive locking of the display lock (which triggers an XCB locking
  assertion), particularly with emacs.

* Add xcb-xlib dependency to x11.pc when built against XCB.

* Allocate the right amount of memory for dpy->lock_fns.  Fixes a crash on
  startup with gdk.


Version 1.1 RC2 - 2006-11-02

Benno Schulenberg:
      nls: remove duplicate compose entries (bug #2286)
      nls: remove shadowed compose entries (bug #2286)
      nls (en_US): remove long compositions that override shorter (bug #2286)

Caolan McNamara:
      XKB geometry: fix leaks in GetKeyboardByName (bug #8099)

David Nusinow:
      Dynamically generate internal manpage section using __libmanpagesuffix__ so that it actually matches the section if you don't use 3X11

Jamey Sharp:
      Add correct Display locking to XKB functions.
      XKB bugfix: SyncHandle must be called after UnlockDisplay, not before.

apps/xclock/xclock.c
/* $Xorg: xclock.c,v 1.4 2001/02/09 02:05:39 xorgcvs Exp $ */
/* $XdotOrg: $ */

/*
 * xclock --  Hacked from Tony Della Fera's much hacked clock program.
 *
 * "-strftime" option added by George Belotsky, Open Light Software Inc.
 */

/*
Copyright 1989, 1998  The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from The Open Group.
 */
/* $XFree86: xclock.c,v 1.16 2002/10/21 13:33:08 alanh Exp $ */


#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <X11/Xatom.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#include "Clock.h"
#include <X11/Xaw/Cardinals.h>
#include "clock.bit"
#include "clmask.bit"

#ifdef XKB
#include <X11/extensions/XKBbells.h>
#endif

#ifndef NO_I18N
#include <locale.h> /* for setlocale() */
Boolean no_locale = True; /* if True, use old behavior */
#endif

#ifdef HAVE_GETPID
# include <unistd.h>
#endif

/* Command line options table.  Only resources are entered here...there is a
   pass over the remaining options after XtParseCommand is let loose. */

static XrmOptionDescRec options[] = {
{"-chime",	"*clock.chime",		XrmoptionNoArg,		"TRUE"},
{"-hd",		"*clock.hands",		XrmoptionSepArg,	NULL},
{"-hands",	"*clock.hands",		XrmoptionSepArg,	NULL},
{"-hl",		"*clock.highlight",	XrmoptionSepArg,	NULL},
{"-highlight",	"*clock.highlight",	XrmoptionSepArg,	NULL},
{"-update",	"*clock.update",	XrmoptionSepArg,	NULL},
{"-padding",	"*clock.padding",	XrmoptionSepArg,	NULL},
{"-d",		"*clock.analog",	XrmoptionNoArg,		"FALSE"},
{"-digital",	"*clock.analog",	XrmoptionNoArg,		"FALSE"},
{"-analog",	"*clock.analog",	XrmoptionNoArg,		"TRUE"},
{"-twelve",	"*clock.twentyfour",	XrmoptionNoArg,		"FALSE"},
{"-twentyfour",	"*clock.twentyfour",	XrmoptionNoArg,		"TRUE"},
{"-brief",	"*clock.brief",		XrmoptionNoArg,		"TRUE"},
{"-utime",	"*clock.utime",		XrmoptionNoArg,		"TRUE"},
{"-strftime",	"*clock.strftime",	XrmoptionSepArg,	NULL},
#ifdef XRENDER
{"-face",	"*clock.face",		XrmoptionSepArg,	NULL},
{"-render",	"*clock.render",	XrmoptionNoArg,		"TRUE"},
{"-norender",	"*clock.render",	XrmoptionNoArg,		"FALSE"},
{"-sharp",	"*clock.sharp",		XrmoptionNoArg,		"TRUE"},
#endif
};

static void quit ( Widget w, XEvent *event, String *params,
		   Cardinal *num_params );

static XtActionsRec xclock_actions[] = {
    { "quit",	quit },
};

static Atom wm_delete_window;

/*
 * Report the syntax for calling xclock.
 */
static void _X_NORETURN
Syntax(const char *call)
{
    fprintf (stderr, "Usage: %s %s", call,
	    "[-analog] [-bw <pixels>] [-digital] [-brief]\n"
	    "       [-utime] [-strftime <fmt-str>]\n"
	    "       [-fg <color>] [-bg <color>] [-hd <color>]\n"
	    "       [-hl <color>] [-bd <color>]\n"
	    "       [-fn <font_name>] [-help] [-padding <pixels>]\n"
	    "       [-rv] [-update <seconds>] [-display displayname]\n"
#ifdef XRENDER
	    "       [-[no]render] [-face <face name>] [-sharp]\n"
#endif
	    "       [-geometry geom] [-twelve] [-twentyfour]\n\n");
    exit(1);
}

static void _X_NORETURN
die(Widget w, XtPointer client_data, XtPointer call_data)
{
    XCloseDisplay(XtDisplayOfObject(w));
    exit(0);
}

static void
quit(Widget w, XEvent *event, String *params, Cardinal *num_params)
{
    Arg arg;

    if (event->type == ClientMessage &&
	(Atom)event->xclient.data.l[0] != wm_delete_window) {
#ifdef XKB
	XkbStdBell(XtDisplay(w), XtWindow(w), 0, XkbBI_MinorError);
#else
	XBell (XtDisplay(w), 0);
#endif
    } else {
	/* resign from the session */
	XtSetArg(arg, XtNjoinSession, False);
	XtSetValues(w, &arg, ONE);
	die(w, NULL, NULL);
    }
}

static void
save(Widget w, XtPointer client_data, XtPointer call_data)
{
    XtCheckpointToken token = (XtCheckpointToken) call_data;
    /* we have nothing to save */
    token->save_success = True;
}

int
main(int argc, char *argv[])
{
    Widget toplevel;
    Arg arg;
    Pixmap icon_pixmap = None;
    XtAppContext app_con;

#ifndef NO_I18N
    char *locale_name = setlocale(LC_ALL,"");
    XtSetLanguageProc ( NULL, NULL, NULL );

    if(!locale_name || 0 == strcmp(locale_name,"C")) {
	no_locale = True;
    }
    else {
	no_locale = False;
    }
#endif

    toplevel = XtOpenApplication(&app_con, "XClock",
				 options, XtNumber(options), &argc, argv, NULL,
				 sessionShellWidgetClass, NULL, ZERO);
    if (argc != 1) Syntax(argv[0]);
    XtAddCallback(toplevel, XtNdieCallback, die, NULL);
    XtAddCallback(toplevel, XtNsaveCallback, save, NULL);
    XtAppAddActions (app_con, xclock_actions, XtNumber(xclock_actions));

    /*
     * This is a hack so that wm_delete_window will do something useful
     * in this single-window application.
     */
    XtOverrideTranslations(toplevel,
		    XtParseTranslationTable ("<Message>WM_PROTOCOLS: quit()"));

    XtSetArg(arg, XtNiconPixmap, &icon_pixmap);
    XtGetValues(toplevel, &arg, ONE);

    if (icon_pixmap == None) {
	arg.value = (XtArgVal)XCreateBitmapFromData(XtDisplay(toplevel),
				       XtScreen(toplevel)->root,
				       (char *)clock_bits, clock_width, clock_height);
	XtSetValues (toplevel, &arg, ONE);
    }

    XtSetArg(arg, XtNiconMask, &icon_pixmap);
    XtGetValues(toplevel, &arg, ONE);
    if (icon_pixmap == None) {
	arg.value = (XtArgVal)XCreateBitmapFromData(XtDisplay(toplevel),
				       XtScreen(toplevel)->root,
				       (char *)clock_mask_bits, clock_mask_width,
				       clock_mask_height);
	XtSetValues (toplevel, &arg, ONE);
    }

    XtCreateManagedWidget ("clock", clockWidgetClass, toplevel, NULL, ZERO);
    XtRealizeWidget (toplevel);
    wm_delete_window = XInternAtom (XtDisplay(toplevel), "WM_DELETE_WINDOW",
				    False);
    (void) XSetWMProtocols (XtDisplay(toplevel), XtWindow(toplevel),
			    &wm_delete_window, 1);

#ifdef HAVE_GETPID
    {
	unsigned long pid = (unsigned long)getpid();
	XChangeProperty(XtDisplay(toplevel), XtWindow(toplevel),
			XInternAtom(XtDisplay(toplevel), "_NET_WM_PID", False),
			XA_CARDINAL, 32, PropModeReplace,
			(unsigned char *) &pid, 1);
    }
#endif

    XtAppMainLoop (app_con);
    exit(0);
}
